### PR TITLE
Ellipsoid related changes and improvements

### PR DESCRIPTION
**Pull request by vaithak**
The changes in the PR have been pulled apart from #171 for easy reviewal and better readability.

### CODE REVIEW

File: include/convex_bodies/ellipsoid.h [hunk 1]
```diff
@@ -28,12 +28,13 @@ typedef Point PointType;
     typedef Eigen::Matrix<NT, Eigen::Dynamic, 1> VT;
     typedef Eigen::Matrix<NT, Eigen::Dynamic, Eigen::Dynamic> MT;
 
+private:
     // representation is (x - c)' A (x - c) <= 1, center is assumed to be origin for now
     MT A;
     Point c;
```
Comment (TolisChal) [inline include/convex_bodies/ellipsoid.h]: In your implementation, the center is always the origin. Do you really need this variable?

File: include/sampling/ellipsoid.hpp [hunk 2]
```diff
@@ -17,23 +17,71 @@
...
+    typedef typename Point::FT NT;
+    typedef typename Eigen::Matrix<NT, Eigen::Dynamic, 1> VT;
+
+    template <typename Ellipsoid, typename RandomNumberGenerator>
+    inline static Point apply(unsigned int const& dim,
+                              Ellipsoid const& E,   // ellipsoid representing the Gaussian distribution
+                              RandomNumberGenerator &rng)
+    {
+        // Generate a point inside a sphere of radius 1.0
+        Point p = GetDirection<Point>::apply(dim ,rng, true);
```
Comment (TolisChal) [inline include/sampling/ellipsoid.hpp]: Plz fix the comma [i.e. ..(dim, rng, true)]

File: include/sampling/ellipsoid.hpp [hunk 3]
```diff
@@ -17,23 +17,71 @@
...
+    typedef typename Eigen::Matrix<NT, Eigen::Dynamic, 1> VT;
+
+    template <typename Ellipsoid, typename RandomNumberGenerator>
+    inline static Point apply(unsigned int const& dim,
+                              Ellipsoid const& E,   // ellipsoid representing the Gaussian distribution
+                              RandomNumberGenerator &rng,
+                              bool normalized=false)
+    {
+        // Generate a point inside a sphere of radius 1.0
+        Point p = GetDirection<Point>::apply(dim ,rng, false);
```
Comment (TolisChal) [inline include/sampling/ellipsoid.hpp]: Plz fix the comma [i.e. ..(dim, rng, true)]

File: include/sampling/ellipsoid.hpp [hunk 4]
```diff
@@ -17,23 +17,71 @@
...
+struct GetGaussianDirection
+{
+    typedef typename Point::FT NT;
+    typedef typename Eigen::Matrix<NT, Eigen::Dynamic, 1> VT;
+
+    template <typename Ellipsoid, typename RandomNumberGenerator>
+    inline static Point apply(unsigned int const& dim,
+                              Ellipsoid const& E,   // ellipsoid representing the Gaussian distribution
+                              RandomNumberGenerator &rng,
+                              bool normalized=false)
```
Comment (TolisChal) [inline include/sampling/ellipsoid.hpp]: I think the boolean `normalized` is redundant here. If we normalize the vector is not Gaussian.

File: include/sampling/ellipsoid.hpp [hunk 5]
```diff
@@ -2,7 +2,7 @@
 
 // Copyright (c) 2012-2020 Vissarion Fisikopoulos
 // Copyright (c) 2018-2020 Apostolos Chalkis
```
Comment (TolisChal) [inline include/sampling/ellipsoid.hpp]: Plz fix the dates

File: include/convex_bodies/ellipsoid.h [hunk 6]
```diff
@@ -28,12 +28,13 @@ typedef Point PointType;
     typedef Eigen::Matrix<NT, Eigen::Dynamic, 1> VT;
     typedef Eigen::Matrix<NT, Eigen::Dynamic, Eigen::Dynamic> MT;
 
+private:
     // representation is (x - c)' A (x - c) <= 1, center is assumed to be origin for now
```
Comment (TolisChal) [inline include/convex_bodies/ellipsoid.h]: c is the origin by default, right?

File: include/convex_bodies/ellipsoid.h [hunk 7]
```diff
@@ -97,9 +112,20 @@ typedef Point PointType;
...
+    // return L_cov * x
+    VT mult_Lcov(VT const& x) const {
+        return _L_cov.template triangularView<Eigen::Lower>() * x;
+    }
+
+
+    void print() const {
         std::cout << "Ellipse is in the form: x' A x <= 1, (center is assumed to be origin always) \n";
-        std::cout << "c = \n" << c.print();
+        std::cout << "c = \n"; c.print();
```
Comment (vissarion) [inline include/convex_bodies/ellipsoid.h]: could you separate in two lines?

**Review by vissarion — APPROVED** [review]

File: include/convex_bodies/ellipsoid.h (hunk 1 above)
Comment (vaithak) [reply include/convex_bodies/ellipsoid.h]: removed the variable 👍🏽

File: include/sampling/ellipsoid.hpp (hunk 3 above)
Comment (vaithak) [reply include/sampling/ellipsoid.hpp]: fixed 👍🏽

File: include/sampling/ellipsoid.hpp (hunk 4 above)
Comment (vaithak) [reply include/sampling/ellipsoid.hpp]: removed 👍🏽

File: include/sampling/ellipsoid.hpp (hunk 5 above)
Comment (vaithak) [reply include/sampling/ellipsoid.hpp]: changed 2020 to 2021 👍🏽

File: include/convex_bodies/ellipsoid.h (hunk 6 above)
Comment (vaithak) [reply include/convex_bodies/ellipsoid.h]: yup, modified the comment.

File: include/convex_bodies/ellipsoid.h (hunk 7 above)
Comment (vaithak) [reply include/convex_bodies/ellipsoid.h]: removed this line, as `c` variable is removed now (as center is always origin)

File: include/sampling/ellipsoid.hpp (hunk 2 above)
Comment (vaithak) [reply include/sampling/ellipsoid.hpp]: fixed 👍🏽

**Review by TolisChal — APPROVED** [review]

**Pull request merged**